File: clients/ruby/stream_wrapper.rb
class StreamWrapper
    @@BYTE_FORMAT_STRING = "c"
    @@INT_FORMAT_STRING = "l<"
    @@LONG_FORMAT_STRING = "q<"
    @@FLOAT_FORMAT_STRING = "e"
    @@DOUBLE_FORMAT_STRING = "E"

    def initialize(stream)
        @stream = stream
    end

    def flush()
        @stream.flush()
    end

    def close()
        @stream.close()
    end

    # Reading primitives

    def read_bool()
        @stream.read_bytes(1).unpack(@@BYTE_FORMAT_STRING)[0] != 0
    end

    def read_int()
        @stream.read_bytes(4).unpack(@@INT_FORMAT_STRING)[0]
    end

    def read_long()
        @stream.read_bytes(8).unpack(@@LONG_FORMAT_STRING)[0]
    end

    def read_float()
        @stream.read_bytes(4).unpack(@@FLOAT_FORMAT_STRING)[0]
    end

    def read_double()
        @stream.read_bytes(8).unpack(@@DOUBLE_FORMAT_STRING)[0]
    end

    def read_string()
        length = read_int()
        @stream.read_bytes(length).pack("U*")
    end

    # Writing primitives

    def write_bool(value)
        @stream.write_bytes([value ? 1 : 0].pack(@@BYTE_FORMAT_STRING))
    end

    def write_int(value)
        @stream.write_bytes([value].pack(@@INT_FORMAT_STRING))
    end

    def write_long(value)
        @stream.write_bytes([value].pack(@@LONG_FORMAT_STRING))
    end

    def write_float(value)
        @stream.write_bytes([value].pack(@@FLOAT_FORMAT_STRING))
    end

    def write_double(value)
        @stream.write_bytes([value].pack(@@DOUBLE_FORMAT_STRING))
    end

    def write_string(value)
        data = value.unpack("U*")
        write_int(data.length())
        @stream.write_bytes(data.pack("C*"))
    end
end

File: clients/python/stream_wrapper.py
import struct


class StreamWrapper:
    BOOL_FORMAT_STRUCT = struct.Struct("?")
    INT_FORMAT_STRUCT = struct.Struct("<i")
    LONG_FORMAT_STRUCT = struct.Struct("<q")
    FLOAT_FORMAT_STRUCT = struct.Struct("<f")
    DOUBLE_FORMAT_STRUCT = struct.Struct("<d")

    def __init__(self, stream):
        self.stream = stream

    def flush(self):
        self.stream.flush()

    def close(self):
        self.stream.close()

    # Reading primitives

    def read_bool(self):
        return self.BOOL_FORMAT_STRUCT.unpack(self.stream.read(1))[0]

    def read_int(self):
        return self.INT_FORMAT_STRUCT.unpack(self.stream.read(4))[0]

    def read_long(self):
        return self.LONG_FORMAT_STRUCT.unpack(self.stream.read(8))[0]

    def read_float(self):
        return self.FLOAT_FORMAT_STRUCT.unpack(self.stream.read(4))[0]

    def read_double(self):
        return self.DOUBLE_FORMAT_STRUCT.unpack(self.stream.read(8))[0]

    def read_string(self):
        length = self.read_int()
        data = self.stream.read(length)
        if len(data) != length:
            raise IOError("Unexpected EOF")
        return data.decode("utf-8")

    # Writing primitives

    def write_bool(self, value):
        self.stream.write(self.BOOL_FORMAT_STRUCT.pack(value))

    def write_int(self, value):
        self.stream.write(self.INT_FORMAT_STRUCT.pack(value))

    def write_long(self, value):
        self.stream.write(self.LONG_FORMAT_STRUCT.pack(value))

    def write_float(self, value):
        self.stream.write(self.FLOAT_FORMAT_STRUCT.pack(value))

    def write_double(self, value):
        self.stream.write(self.DOUBLE_FORMAT_STRUCT.pack(value))

    def write_string(self, value):
        data = value.encode("utf-8")
        self.write_int(len(data))
        self.stream.write(data)
